### PR TITLE
Add nested key getters `get_if` and `get_or` for record

### PR DESCRIPTION
**Pull request by jachris**
This PR does not yet refactor the existing code to make use of this.

Closes tenzir/issues#170.

### CODE REVIEW

File: libvast/test/data.cpp [hunk 1]
```diff
@@ -404,3 +404,43 @@ TEST(pack / unpack) {
...
 }
+
+TEST(get_if) {
+  // clang-format off
+  auto x = record{
+    {"foo", "bar"},
+    {"baz", record{
+      {"qux", int64_t{42}},
+      {"quux", record{
+        {"quuux", 3.14}
```
Comment (dominiklohmann) [inline libvast/test/data.cpp:415]: NB: If you add a trailing comma here it should format the same way even with clang-format enabled for the section.

File: libvast/test/data.cpp [hunk 2]
```diff
@@ -404,3 +404,43 @@ TEST(pack / unpack) {
...
+  auto unknown = get_if<ip>(&x, "foo.baz");
+  CHECK(!unknown);
+}
+
+TEST(get_or) {
+  auto x = record{{"foo", "bar"}};
+  auto fallback = std::string{"fallback"};
+  auto foo = get_or(x, "foo", fallback);
+  CHECK_EQUAL(foo, "bar");
+  auto bar = get_or(x, "bar", fallback);
```
Comment (dominiklohmann) [inline libvast/test/data.cpp:444]: Can you also add a test for get_or with a fallback string literal? I think there's a missing overload.

File: libvast/include/vast/data.hpp [hunk 3]
```diff
@@ -249,6 +251,73 @@ void merge(const record& src, record& dst,
...
+    }
+  }
+  die("unreachable");
+}
+
+template <typename T>
+T* get_if(record* r, std::string_view path) {
+  auto result = get_if<T>(static_cast<const record*>(r), path);
+  return const_cast<T*>(result); // NOLINT
+}
```
Comment (dominiklohmann) [inline libvast/include/vast/data.hpp:304]: Small scope creep request: Can you also add an unchecked `T& get(record&, std::string_view)` (const and non-const) that just calls `get_if`, asserts, and then dereferences?

File: libvast/include/vast/data.hpp [hunk 4]
```diff
@@ -249,6 +251,73 @@ void merge(const record& src, record& dst,
...
+      auto f = detail::overload{
+        [](const T& result) {
+          return &result;
+        },
+        [](const auto&) -> const T* {
+          // Found, but type does not match.
+          return nullptr;
+        },
+      };
+      return caf::visit(f, field);
```
Comment (dominiklohmann) [inline libvast/include/vast/data.hpp:281]: You could actually use `if (const auto* value = caf::get_if<T>(field)) { ... }` here to save a few lines and a few template instantiations, but I don't mind either way. The current form is more easily extensible if we ever want to support casting here.

File: libvast/include/vast/data.hpp [hunk 5]
```diff
@@ -249,6 +251,73 @@ void merge(const record& src, record& dst,
...
+    }
+    auto f = detail::overload{
+      [](const record& sub_record) {
+        return &sub_record;
+      },
+      [](const auto&) -> const record* {
+        return nullptr;
+      },
+    };
+    current = caf::visit(f, field);
```
Comment (dominiklohmann) [inline libvast/include/vast/data.hpp:291]: Same as above, could be a single `caf::get_if`.

**Review by dominiklohmann — APPROVED** [review]
This is some really great code. Approving as-is, with some minor comments inline.

File: libvast/test/data.cpp (hunk 1 above)
Comment (jachris) [reply libvast/test/data.cpp:415]: I don't see a way to get the same formatting. clang-format always puts `record` on a separate line.

Comment (dominiklohmann) [reply libvast/test/data.cpp:415]: Right, I see what you mean. The result is then this:

```cpp
auto x = record{
  {"foo", "bar"},
  {"baz",
   record{
     {"qux", int64_t{42}},
     {"quux",
      record{
        {"quuux", 3.14},
      }},
   }},
};
```

I don't have a strong enough opinion on this to force a personal preference to avoid disabling clang-format, so feel free to just resolve.

File: libvast/test/data.cpp (hunk 2 above)
Comment (jachris) [reply libvast/test/data.cpp:444]: Ah, right, I missed the convenience overload. We could additionally provide an overload for cheap-to-copy types. Thoughts?

Comment (dominiklohmann) [reply libvast/test/data.cpp:444]: We agreed that we want to do this in a follow-up if we need it. Resolving as-is with the new string_view overload.

**Review by dominiklohmann — APPROVED** [review]
